### ISSUE TITLE
Physics simulation stops when generated convex hulls collide

### ISSUE THREAD

**Issue by isaac-mason** [donmccurdy/three-to-cannon#76]
It looks like there may be some issues with generated convex hulls. When two bodies with generated convex hull shapes collide, the physics simulation stops and the page becomes unresponsive.

I made a simple codesandbox demonstrating the issue: https://codesandbox.io/s/repro-three-to-cannon-hull-collision-issue-68jtcd?file=/src/index.ts

I haven't checked what versions have the issue, but it appears to at least be present in 4.0.2, predating the introduction of the `getShapeParameters` function. The issue may be somewhere in [ConvexHull.js](https://github.com/donmccurdy/three-to-cannon/blob/main/lib/ConvexHull.js)?

**Comment by isaac-mason** [comment]
Happy to look into this too - just need to find some time 🙂

**Comment by donmccurdy** [comment]
The THREE.ConvexHull is not used by the simulation at runtime — only to construct a [CANNON.ConvexPolyhedron](https://pmndrs.github.io/cannon-es/docs/classes/ConvexPolyhedron.html). I would guess that because collisions with other types of shapes still work, and the debug outlines look reasonable, the hulls themselves are valid. The way this hangs the browser, it looks a bit like an infinite loop in collision code?

**Comment by isaac-mason** [comment]
Looks like this might be a duplicate of this issue: https://github.com/donmccurdy/three-to-cannon/issues/59

**Comment by isaac-mason** [comment]
Potentially relevant: https://github.com/schteppe/cannon.js/issues/459#issuecomment-887939884

**Comment by donmccurdy** [comment]
The simplifyGeometry() method is intended to ensure vertices are merged and convex triangles are connected...

https://github.com/donmccurdy/three-to-cannon/blob/2927293f04f1551d16723e0ec15d4b32c4f1c6ac/src/utils.ts#L114-L122

...but we invoke it only if we're merging multiple geometries —

https://github.com/donmccurdy/three-to-cannon/blob/2927293f04f1551d16723e0ec15d4b32c4f1c6ac/src/utils.ts#L17-L29

This might be the cause? Unfortunately it's not as simple as just adding a call to `simplifyGeometry` above — if we simplify single geometries then we can't get precise parameters for primitive geometries, by accessing things like `geometry.parameters.radiusTop` for cylinders.

There's a lot in this library that could use serious refactoring and better unit tests, sorry for that. 😮‍💨 If you'd be interested in push access and just making more changes as you see fit, I'm happy to go that direction.

**Comment by isaac-mason** [comment]
Thanks @donmccurdy for the info! I think it's sent me down a good path 🙂 

It looks like the issue may be in how `getConvexPolyhedronParameters` constructs faces? Reading [`ConvexPolyhedron.clipFaceAgainstHull`](https://github.com/pmndrs/cannon-es/blob/master/src/shapes/ConvexPolyhedron.ts#L472) in the cannon-es repo makes the issue a little clearer.

Heres my current understanding:

As part of convex-convex collision, connected faces are built for a given hull face. Connected faces are retrieved by searching for face vertices in `convexPolyhedron.faces` that are shared with a given faces vertices e.g. `polyA.indexOf(hullA.faces[i][j]) !== -1`.

In [`getConvexPolyhedronParameters`](https://github.com/donmccurdy/three-to-cannon/blob/6f15fd2/src/index.ts#L274
), `currentFaceVertex` is incremented for each face. So when cannon-es `ConvexPolyhedron.clipFaceAgainstHull` is looking for connected faces, it won't find any, as all of the face vertex values produced by `getConvexPolyhedronParameters` are unique.

**Comment by isaac-mason** [comment]
I'm pretty keen to get to the bottom of generating convex hulls that support convex-convex collision! Happy to keep on with this and (hopefully) create a PR once I've got something working.

From there, I'd also be happy to help with general improvements/upkeep, for sure 🙂 

> There's a lot in this library that could use serious refactoring and better unit tests, sorry for that. 😮‍💨 If you'd be interested in push access and just making more changes as you see fit, I'm happy to go that direction.

No apologies are necessary! Thanks for creating and supporting this lib!

**Comment by isaac-mason** [comment]
Sorry, haven't had much time to progress this.

I did however notice there's a new PR in the cannon-es repo for simplifying the creation of convex hulls: https://github.com/pmndrs/cannon-es/pull/155

I did a quick test, simply pulled the QuickHull class from that PR into three-to-cannon, and convex-convex collision works with the faces it returns.

I'm going to watch that PR for now and see if something can be merged into cannon-es itself. If not, pulling that implementation into this lib could be good too!
